### PR TITLE
Create launch mode-specific classpath configurations for Gradle applications

### LINKED ISSUE #13052: Gradle quarkusDev fails with multi-module runtimeOnly dependencies
**Describe the bug**
I have a project with multiple modules. The app module contains the quarkus plugin and has a runtimeOnly dependency on moduletwo.

Running `./gradlew clean quarkusDev` results in a stacktrace

```
Execution failed for task ':app:quarkusGenerateCode'.
> Error while reading file as JAR: /path/quarkus-test/moduletwo/build/libs/moduletwo-1.0.0-SNAPSHOT.jar
```

**Expected behavior**
Quarkus Plugin should ensure that other modules within the project have been compiled before attempting to depend on them

**Actual behavior**
Quarkus Gradle plugin does not seem to have compileJava dependency on internal runtimeOnly modules. At the moment there are two workarounds:

1. Using implementation instead of runtimeOnly in module dependencies
2. Manually ensuring that the module is built. e.g
```
./gradlew build
./gradlew quarkusDev
```

**To Reproduce**

Link to a small reproducer (preferably a Maven project if the issue is not Gradle-specific).

https://github.com/huw0/quarkus-multimodule-runtimeOnly

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: N/A
 - Output of `java -version`: 11.0.9
 - GraalVM version (if different from Java): N/A not native build
 - Quarkus version or git rev: 1.9.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Gradle 6.6.1

### PR DESCRIPTION
**Pull request by aloubyansky**
This is a refactoring of the classpath configuration initialization and resolution for Gradle projects.
It is meant to make it more clear how classpath configurations are initialized and resolved for each launch mode using dedicated configuration names for each (they should probably be turned into constants instead of being dynamically composed). It also removes adding deployment modules as variants.

It's still not as clean as I would like it to be but, imo, it is a bit better and also allows us to implement commands that could force certain dependencies on the classpath. Such as container image creating commands @iocanel is prototyping.

Fix #13052 in the second commit.

### CODE REVIEW

File: devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                .extendsFrom(configContainer.getByName(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME));
+
+        configContainer.create(ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(LaunchMode.NORMAL))
+                .extendsFrom(configContainer.getByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME));
+
+        configContainer.create(ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(LaunchMode.DEVELOPMENT))
+                .extendsFrom(
+                        configContainer.getByName(ToolingUtils.DEV_MODE_CONFIGURATION_NAME),
+                        configContainer.getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME),
+                        configContainer.getByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME));
```
Comment (glefloch) [inline devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java:70]: shouldn't we also extend from `TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME ` for continuous testing?

**Review by glefloch — COMMENTED** [review]
Nice changes, I just left a few comments.

File: devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java (hunk 1 above)
Comment (aloubyansky) [reply devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java:70]: No, it's a different launch mode. We are resolving a separate `ApplicationModel` for each one.

Comment (glefloch) [reply devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java:70]: But when running `gradle quarkusDev` we only resolve one model right?

Comment (aloubyansky) [reply devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java:70]: No, we resolve both. https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java#L331

**Review by glefloch — APPROVED** [review]

**Pull request merged**